src/windows/models/properties_model.py

```diff
@@ -447,51 +447,52 @@ def value_updated(self, item, interpolation=-1, value=None, interpolation_detail
                     # Loop through points, find a matching points on this frame
                     found_point = False
                     point_to_delete = None
-                    for point in clip_data[property_key]["Points"]:
```

Comment:
I'm on my phone and can't do proper suggestions right now, but this should just be changed to:
```python
for point in clip_data[property_key].get("Points", []):
```
That will protect it without adding an additional level of nesting. If `"Points"` isn't a key, an empty list will be returned and the loop skipped.

Reply:
Ugh. And as the comment I just LOST was saying: neither your change or mine will be safe if there's any possibility that `clip_data[property_key]` isn't a `dict` AT ALL. If that's a possibility, we probably need more sanity-checking earlier in the code.

Reply:
The error on Sentry is listed as `KeyError 'Points'`. So I'm thinking this is a `dict`, but just does not always contain a Points keys. At least, that is what it appears so far.

Reply:
@jonoomph That makes sense, so get("Points", []) would prevent that on this line.

But on line 506, we append to 'Points' and that will be a problem if 'Points' isn't an attribute of `clip_data[property_key]`

[Here](https://stackoverflow.com/questions/2052111/efficient-way-to-either-create-a-list-or-append-to-it-if-one-already-exists) I found this solution:
```python
clip_data[property_key].setdefault('Points', []).append(<point>)
```
I think it's pretty robust, because it will create, or append to an existing list.

Together, these let us deal with clips having no points.

Reply:
Yup, that should work nicely.

Reply:
If we're seeing property entries that _are_ dicts, but lack `'Points'` keys, then that probably means users are deleting the last keyframe point on their properties. IIRC that could cause this. (It certainly causes _other_ problems!)

There's code in my keyframe-processing rewrite (#3840) that refused to do that even if the user asked to. But as I suspect that PR is too diverged from `develop` to ever be mergeable, the last-keyframe protection should probably get ported to the existing code.

Reply:
> There's code in my keyframe-processing rewrite (#3840) that refused to do that even if the user asked to. But as I suspect that PR is too diverged from `develop` to ever be mergeable

Actually, maybe not! I just finished rebasing it on `develop`, included fixes for the issues here, and even found a solution for the data-optimization bug that crept in with the object_id changes a while back. (A fix which I can also apply to `develop`. I'll create a PR now for that change.)

Reply:
Done, in #4213. Any and _all_ review welcome. It's **immensely** pared down from #3840, but still a lot to review. The only thing I can suggest is to review it commit-by-commit, instead of all the changes at once. I broke it into a bunch of commits, because each one on its own is pretty self-contained and readable (relatively), even if the full diff is a nightmare. (Primarily because of a bunch of un-indenting I did, thanks to some newly-added early returns.)

Reply:
Taking a look. And thanks for breaking it up!